Quote & List v2: Deleting empty list item should list block

ellatrix opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

Fixes #40979.

## Why?

It's not possible to delete an empty quote or list block in one go.

## How?

A generalised solution: When deleting all inner blocks, check if the block is otherwise empty (unmodified) and delete the whole block if so.

## Testing Instructions

Create a quote block with a paragraph. Delete the paragraph through the UI. The quote block should be deleted too.

## Screenshots or screencast <!-- if applicable -->

ntsekouras commented on packages/e2e-tests/specs/editor/blocks/quote.test.js:
```diff
@@ -147,4 +147,15 @@ describe( 'Quote', () => {
+		await page.keyboard.type( 'b' );
+
+		// Expect "ab" in the paragraph.
+		expect( await getEditedPostContent() ).toMatchSnapshot();
```
Nit: let's use `toMatchInlineSnapshot` here. It's a small snapshot.

ntsekouras commented on packages/block-editor/src/store/actions.js:
```diff
@@ -1165,9 +1165,22 @@ export const removeBlocks =
+		const { __unstableRemoveEmpty } =
+			select.getBlockListSettings( rootClientId ) || {};
+
+		if ( blockOrder.length === 0 && __unstableRemoveEmpty ) {
```
If we go this way, I guess we should make the checks and dispatch once for removal and selection. Also `dispatch( ensureDefaultBlock() );` seems to check internally the `count` so we could leave without a check..

ellatrix replied:
Fixed.

ntsekouras reviewed (commented):
Thanks for the PR Ella! Interesting stuff here and the API changes should be highlighted to the title/description of the PR.

I'll have to process it more, but I'm quite conflicted on this one..

I know when we remove all innerBlocks(ex. Group) is not in a great state right now, but I think most of the discussions around that were about improving that state/design. Of course this doesn't mean we shouldn't consider the new flag. I think what felt weird to me is when I removed the last block from the toolbar actions and removed everything..

Also in order for this to take effect in RichText blocks we need to have an `onRemove` property there too and that can make some experience not feel consistent? 🤔 . 

In my video I added the flag in `Group`(I guess I could use Quote now too 😄 ) and doing what I'm describing above. For example Site Title doesn't have an `onRemove`.

https://user-images.githubusercontent.com/16275880/180016905-36b00a79-b1d8-4082-aada-192fb55a4684.mov

As I said I need to process this more to form a better opinion and would love to hear what others think.

ellatrix commented:
> I know when we remove all innerBlocks(ex. Group) is not in a great state right now, but I think most of the discussions around that were about improving that state/design.

Is this an argument against this change? I'm not sure I'm following. :)

> I think what felt weird to me is when I removed the last block from the toolbar actions and removed everything..

Why does it feel weird? If you remove the last list item, you also remove the list. Is that unexpected?

> Also in order for this to take effect in RichText blocks we need to have an onRemove property there too and that can make some experience not feel consistent? 🤔 .

That's just for removing the block with Backspace, which is independent from removing the wrapper block when all inner blocks have been removed.

> In my video I added the flag in Group(I guess I could use Quote now too 😄 ) and doing what I'm describing above. For example Site Title doesn't have an onRemove.

Well, then it's also not possible to remove Site Title anywhere else like that? This is a problem with Site Title, not the group block. Just try Site Title outside of a group block and you have the same problem.

ntsekouras commented:
>Is this an argument against this change? I'm not sure I'm following. :)

I'm not against it's just this behavior was mostly discussed about Group block and I have the feeling that folks were leaning on improving the design when we have no blocks. This doesn't mean we can't have this new API.

>Why does it feel weird? If you remove the last list item, you also remove the list. Is that unexpected?

Maybe a bit, but that's subjective and no strong opinion on this one 😄 

I'll re-test the PR.

ntsekouras commented on packages/block-editor/src/store/actions.js:
```diff
@@ -1159,14 +1159,27 @@ export const removeBlocks =
 
-		// To avoid a focus loss when removing the last block, assure there is
-		// always a default block if the last of the blocks have been removed.
+		if ( removeEmpty ) {
```
We should also combine this with something like:
```
dispatch( {
	type: 'REMOVE_BLOCKS',
	clientIds: removeEmpty ? [ rootClientId ] : clientIds,
} );
```
to create only one history step.

ellatrix commented:
It's what we're doing now for the List v1. When you delete the only list item, it removes the block. Same with Quote v1.
Doing this for List and Quote doesn't mean all blocks should do this. Group can remain as it is. That's why it's an option. :)

ntsekouras reviewed (commented):
The only thing with this approach that we need to consider is the case where a block has other things besides innerBlocks. The example here would be the `Quote` block when we have set a `citation`. If we remove the last innerBlock it would still remove the whole block.

Would that be expected or okay? If yes, it should be documented somewhere what we consider `empty` in comments.

ellatrix commented:
@ntsekouras Let's put this on hold. There may indeed be better ways to do this for keyboard. But I suspect this PR might still be necessary to handle https://github.com/WordPress/gutenberg/issues/40979. In any case, this is lower priority then. I'm making an alternative PR for keyboard.

t-hamano commented:
This PR may also resolve #45919.
Please see the comment [here](https://github.com/WordPress/gutenberg/issues/45919#issuecomment-1326492758).

ellatrix committed:
General fix

jorgefilipecosta reviewed (approved):
I think we should give it a try to merge this PR. It solves so many issues e.g: button block (removing the last button should remove the buttons block itself, list, media & text). In the default blocks core support, this behavior seems to be positive in the user experience.  I think for most third-party blocks this behavior is also positive, I can not think easily of a case where a block has inner blocks and if we don't change anything on the block and remove the inner blocks we want the block to still be there. But a case we are not considering may exist, I guess adding a dev note for this change could be useful to raise awareness on third-party block developers, I will leave the decision on adding the needs dev note label to @ellatrix.

The change itself looks good and worked well on my tests 👍

colorful-tones commented:
Hi folks,

Any chance the release team could please get an update on any blockers or the potential for this to be a part of the WordPress 6.5 release? Today is the 6.5 Beta 1 release, and I'm not sure if this is a critical piece. Would love a brief update and if we should triage this off the [WordPress 6.5 Editor Tasks board](https://github.com/orgs/WordPress/projects/148), or other?

colorful-tones commented:
Based on a recent Editor Triage async session - I'm moving this to the **Punted to 6.6** column on the [WordPress 6.5 Editor Tasks board](https://github.com/orgs/WordPress/projects/148/views/1).

colorful-tones commented:
FWIW - I just pulled this down and rebased with `trunk`.

I then tested and obtained the same results. Perhaps this is already fixed and can be closed as it is no longer an issue?

**WP 6.6 Beta-2 + Gutenberg 18.5**

https://github.com/WordPress/gutenberg/assets/405912/9ae49893-4f73-4c70-b6dd-23650de0d7d7

**WP 6.6 Beta-2 + this PR rebased**

https://github.com/WordPress/gutenberg/assets/405912/cd7660e5-d854-45de-823e-9795ed00b1be

t-hamano commented:
> Perhaps this is already fixed and can be closed as it is no longer an issue?

Regarding keyboard actions, they work fine in trunk, but the issue persists when removing blocks from the block toolbar. I think this PR will fix that remaining issue.

After rebasing this PR, I think it would be ok to backport it to WP6.6, but what do you think?

### trunk

When the only block in a List block or  Quote block is deleted from the block toolbar, an empty block with no inner blocks remains.


https://github.com/WordPress/gutenberg/assets/54422211/8bdd0264-1a58-4602-96d9-9d7bb28e3de3

### This PR

The issue in trunk will be resolved.

Note: This PR depends on `rootClientId`, but the `rootClientId` variable was deleted in #62547, so an undefined variable error occurs. Therefore, in order to properly test this PR, you need to rebase this PR with the latest trunk and then apply the following changes:

```diff
diff --git a/packages/block-editor/src/store/private-actions.js b/packages/block-editor/src/store/private-actions.js
index fc2a28a75f..27b276f1e3 100644
--- a/packages/block-editor/src/store/private-actions.js
+++ b/packages/block-editor/src/store/private-actions.js
@@ -154,6 +154,7 @@ export const privateRemoveBlocks =
                        }
                }
 
+               const rootClientId = select.getBlockRootClientId( clientIds[ 0 ] );
                const blockOrder = select.getBlockOrder( rootClientId );
                const removeRoot =
                        blockOrder.length === clientIds.length &&
```

https://github.com/WordPress/gutenberg/assets/54422211/445f8502-f5b1-4df6-802d-996f839c0a2a

ellatrix commented:
I think @ntsekouras had a good point about "what if there's other content besides inner blocks".
Maybe we should just disable the remove button, if it's required by the template to have at least 1?

t-hamano commented:
> if it's required by the template to have at least 1?

How can we determine that a block needs at least one inner block? Does it require any new block support 🤔

ellatrix commented:
We can check the inner block template

jeryj commented:
To be consistent, should deleting the last image of a gallery block remove the entire gallery block? Right now it reverts back to the gallery placeholder.

t-hamano commented:
> Right now it reverts back to the gallery placeholder.

I believe this is because the `linTo` attribute is updated when the block is mounted.

https://github.com/WordPress/gutenberg/blob/ec35342e1c1e46232a3628e0d4d81fbbc43911dd/packages/block-library/src/gallery/edit.js#L593-L603

We should be able to resolve this by adding `'content'` as the second argument to `isUnmodifiedBlock`.

https://github.com/WordPress/gutenberg/blob/ec35342e1c1e46232a3628e0d4d81fbbc43911dd/packages/block-editor/src/store/private-actions.js#L186-L188

However, this alone does not solve all the problems. For example, even if you delete the last child block of a group block, the group block itself is not deleted. Perhaps we need to check not only for modified blocks but also for modified block variations.
